### PR TITLE
Add priority_weight to ti_trigger_id idx

### LINKED ISSUE #40674: Expensive Trigger Query Does Not Use Index (even after forcing a hint) Due to Additional Column
### Apache Airflow version

2.9.2

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

We are planning to upgrade from 2.7.3 to 2.9.2. However, we've observed issues in our sandbox MySQL db while doing so. Specifically, [this query](https://github.com/apache/airflow/blob/main/airflow/models/trigger.py#L306-L314) fails to use the ti_trigger_id index on the task_instance table. This is likely a result of the addition of the `coalesce(TaskInstance.priority_weight, 0).desc()` which forces MySQL to need to perform lookups after the fact and decide that the index is not the most optimal route.

Slow Query Log (showing that this query takes over 11s to run):

```
# Query_time: 11.081314  Lock_time: 0.000299 Rows_sent: 0  Rows_examined: 2819837
SET timestamp=1720544726;
SELECT `trigger`.id 
FROM `trigger` INNER JOIN task_instance ON `trigger`.id = task_instance.trigger_id 
WHERE `trigger`.triggerer_id IS NULL OR (`trigger`.triggerer_id NOT IN (SELECT job.id 
FROM job 
WHERE job.end_date IS NULL AND job.latest_heartbeat > '2024-07-09 17:04:56.519611' AND job.job_type = 'TriggererJob')) ORDER BY coalesce(task_instance.priority_weight, 0) DESC, `trigger`.created_date 
 LIMIT 9968 FOR UPDATE SKIP LOCKED;
```

### What you think should happen instead?

The index on trigger_id should be modified to include priority_weight so that the query chooses to use the index. Currently, the index is defined as 

```
CREATE INDEX ti_trigger_id
   ON task_instance (trigger_id ASC);
```

However, we believe it should be defined as the following so that the index gets used:

```
CREATE INDEX idx_trigger_priority ON task_instance(trigger_id, priority_weight);
```

### How to reproduce

Run Airflow w/ a MySQL backend and turn on slow query logging. Airflow should be bootstrapped with a signficant number of tasks in the task_instance table

### Operating System

Debian 11

### Versions of Apache Airflow Providers

https://raw.githubusercontent.com/apache/airflow/constraints-2.9.2/constraints-3.11.txt

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

KubernetesExecutor on GKE

### Anything else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by vchiapaikeo**
closes: https://github.com/apache/airflow/issues/40674

This is my first db migration change so not exactly sure how to test it. If you have any tips, please let me know!

I've started breeze and ensured that migrations were able to run. Also, I've ensured that the triggerer process is able to run successfully. 

```
Checking backend and integrations.

PostgreSQL: OK.


Starting Airflow

DB: postgresql+psycopg2://postgres:***@postgres/airflow
Performing upgrade to the metadata database postgresql+psycopg2://postgres:***@postgres/airflow
[2024-07-10T14:20:24.200+0000] {migration.py:215} INFO - Context impl PostgresqlImpl.
[2024-07-10T14:20:24.200+0000] {migration.py:218} INFO - Will assume transactional DDL.
[2024-07-10T14:20:24.201+0000] {migration.py:215} INFO - Context impl PostgresqlImpl.
[2024-07-10T14:20:24.201+0000] {migration.py:218} INFO - Will assume transactional DDL.
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> d482b7261ff9
Database migrating done!
/usr/local/lib/python3.10/site-packages/flask_limiter/extension.py:333 UserWarning: Using the in-memory storage for tracking rate limits as no storage was explicitly specified. This is not recommended for production use. See: https://flask-limiter.readthedocs.io#configuring-a-storage-backend for documentation about configuring the storage backend.
[2024-07-10T14:20:27.328+0000] {override.py:1454} INFO - Inserted Role: Admin
[2024-07-10T14:20:27.330+0000] {override.py:1454} INFO - Inserted Role: Public
```

image

image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by vchiapaikeo** [comment]
There is also a FK defined around this -->

```
        ForeignKeyConstraint(
            [trigger_id],
            ["trigger.id"],
            name="task_instance_trigger_id_fkey",
            ondelete="CASCADE",
        ),
```

Unsure if we need to drop the constraint before performing the migration. I was able to do this on my database without needing to drop the constraint.

**Comment by potiuk** [comment]
I think this is yet another case where MySQL has to be approached differently - generally our tests also run migrations up-> down an if you look there MySQL migrations failed (precisely because of the constraints). This is the power of our CI ;).

I am not sure how to fix it though - I've been trying to keep away from those migrations - but there are likely other similar migrations already so probably it's a matter of finding out the pattern that is used there.

**Comment by vchiapaikeo** [comment]
Yesss, the CI/CD is 💯 . Going to drop constraints as well. This is a noop for PG but is required for MySQL so I ended up not conditioning it and doing the constraint drop / recreate for both.

**Comment by potiuk** [comment]
Does it actually fix the problem BTW? In the linked issue you had some doubts?

**Comment by vchiapaikeo** [comment]
So I tried to fix [the query](https://github.com/apache/airflow/blob/main/airflow/models/trigger.py#L306-L314) (which I thought would be a better, simpler solution) to use a left join but actually, Postgres does not allow row locks with a left join.

On our sandbox MySQL instance where we have ~3M task_instances, this is producing the correct explain plan now. 

Full table scan on trigger table instead of task_instance table:

airflow-db

Previously, it was doing the full table scan on the task_instance table:

airflow-db-prev

**Comment by potiuk** [comment]
LGTM. Let me just summon @dstandish and @ephraimbuddy to counter-verify -> those are the two experts I know from the area and we had recently discussion that adding indexes should be reviewed for more than one pair of eyes.

**Review by potiuk — APPROVED** [review]

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1797,7 +1797,7 @@ class TaskInstance(Base, LoggingMixin):
         Index("ti_state_lkp", dag_id, task_id, run_id, state),
         Index("ti_pool", pool, state, priority_weight),
         Index("ti_job_id", job_id),
-        Index("ti_trigger_id", trigger_id),
+        Index("ti_trigger_id", trigger_id, priority_weight),
```
Comment (potiuk) [inline airflow/models/taskinstance.py:1800]: Just commenting so that it's not merged accidentally before others look at it .

**Review by eladkal — APPROVED** [review]

**Pull request closed without merging**